### DISCUSSION TITLE
Как обеспечить "чистоту" обработки Действий со случайным исходом?

### DISCUSSION THREAD

**Discussion by nergal-perm** [nergal-perm/tg-core#26] [Закрытые обсуждения] [answered]
Как обеспечить однозначную повторяемость действий при работе со случайными величинами вроде порядка карт в перемешанной колоде? 

Допустим, на вход Игре подаются две Команды: _Перемешать Колоду_ и _Взять первую Карту из Колоды_. Каким образом в тестах или при повторном проигрывании действий обеспечить одинаковый порядок карт в Колоде и таким образом гарантировать, что второе действие в этой последовательности будет **всегда** иметь один и тот же результат? 

При этом, нужно учесть, что в реальной Игре результат второго действия должен **всегда** быть случайным...

**Comment by nergal-perm** [comment]
Есть вариант задавать "режим" игры, что-то вроде "интерактивный" и "повтор". В первом случае Команды, создающие случайные данные, реально перемешивают Колоды и бросают Кубики, а во втором - используют поддельные реализации с заранее заданными исходами. По идее, в "интерактивном" режиме можно использовать специального Секретаря, который бы сохранял состояние случайных величин для последующего одинакового проигрывания партий в режиме "повтор".

**Reply by citerm** [reply]
я думаю если мы сделаем крупье (перемешивающего карты) отдельным сервисом или модулем, то он будет передавать в движок действие с конкретным объектом перемешанной колоды, и этот объект можно взять из лога, и передавать повторно

**Reply by citerm** [reply]
**UPD**
Для системы результат работы такого модуля-крупье не будет случайным исходом, это будет действие перемешивания колоды с определенными параметрами колоды. То же самое с кубиком, мешком и другими элементами игры, которые генерируют псевдослучайные структуры данных. Даже если этот модуль будет вызываться в функции-действии, мы всегда можем описать результат действия для каждого выпавшего результата, если результат действительно влияет на состояние.
Но на мой взгляд действие игрока "сделать ход, бросив кубик", должно быть разбито на несколько действий:

- бросить кубик (итогом действия является ограничение текущего игрока делать другие действия, пока не получен результат кубика - действие "бросить" вызывает функцию рассчитать результат кубика
- рассчитать результат кубика (это действие по запросу игры выполняет специальный модуль-крупье, которому в данный момент разрешено выполнять действия с игрой - он асинхронно рассчитывает результат кубика и передает в единый обработчик действий результат кубика, который меняет состояние игры, и на этом ход игрока завершается)

По аналогии как мы рассуждали с шахматами. При таком подходе мы сможем эмулировать работу модуля-крупье, вбрасывая результаты расчетов псевдослучайных структур самим через единый канал обработки действий.

**Comment by nergal-perm** [comment, marked as answer]
Записывать для последующего проигрывания мы всегда будем уже сформированный результат случайного исхода. То есть, в истории игры появится запись вроде "Игрок выкинул 5 и 4 на кубиках" или "Игрок вытянул семерку треф из колоды".

Подробнее в #60 и [ADR Сохранять изменения состояния системы в виде событий](https://github.com/nergal-perm/tg-core/wiki/0002-persist-state-changes-as-events)
